encoder.lua

```diff
@@ -193,55 +348,99 @@ local mpv = { }
+        }
+    else
+        encoder_args = {
+            '--ovc=mjpeg',
+            '--vf-add=scale=out_range=jpeg',
+            string.format(
+                    '--ovcopts=global_quality=%d*QP2LAMBDA,flags=+qscale',
+                    quality_to_crf_jpeg(self.config.snapshot_quality)
+            ),
+        }
     end
-    return args
+
+    return mpv.prepend_common_args(
+            source_path,
+            '--audio=no',
+            '--frames=1',
+            '--start=' .. toms(timestamp),
+            mpv.get_scale_arg("sinc", self.config.snapshot_width, self.config.snapshot_height),
+            '-o=' .. output_path,
+            h.unpack(encoder_args)
+    )
 end
 
 mpv.make_animated_snapshot_args = function(source_path, output_path, start_timestamp, end_timestamp)
-    return {
-        mpv.exec,
-        source_path,
-        '--loop-file=no',
-        '--keep-open=no',
-        '--ovc=libwebp',
-        '--of=webp',
-        '--ofopts-add=loop=0',
-        '--audio=no',
-        '--no-sub',
-        '--no-ocopy-metadata',
-        '--ovcopts-add=lossless=0',
-        '--ovcopts-add=compression_level=6',
-        table.concat { '--start=', toms(start_timestamp), },
-        table.concat { '--end=', toms(end_timestamp), },
-        table.concat { '--ovcopts-add=quality=', tostring(self.config.animated_snapshot_quality) },
-        table.concat { '--vf-add=scale=', self.config.animated_snapshot_width, ':', self.config.animated_snapshot_height, ':flags=lanczos', },
-        table.concat { '--vf-add=fps=', self.config.animated_snapshot_fps, },
-        table.concat { '-o=', output_path },
-    }
+    local encoder_args
+    if self.config.animated_snapshot_format == 'avif' then
+        encoder_args = {
+            '--ovc=libaom-av1',
+            -- cpu-used < 6 can take a lot of time to encode.
+            '--ovcopts-add=cpu-used=6',
+            string.format('--ovcopts-add=crf=%d', quality_to_crf_avif(self.config.animated_snapshot_quality)),
+        }
+    else
+        encoder_args = {
+            '--ovc=libwebp',
+            '--ovcopts-add=compression_level=6',
+            string.format('--ovcopts-add=quality=%d', self.config.animated_snapshot_quality),
+        }
+    end
+
+    return mpv.prepend_common_args(
+            source_path,
+            '--audio=no',
+            '--start=' .. toms(start_timestamp),
+            '--end=' .. toms(end_timestamp),
+            '--ofopts-add=loop=0',
+            string.format('--vf-add=fps=%d', self.config.animated_snapshot_fps),
+            mpv.get_scale_arg("lanczos", self.config.animated_snapshot_width, self.config.animated_snapshot_height),
+            '-o=' .. output_path,
+            h.unpack(encoder_args)
+    )
 end
 
-mpv.make_audio_args = function(source_path, output_path, start_timestamp, end_timestamp)
+mpv.make_audio_args = function(source_path, output_path,
+                               start_timestamp, end_timestamp, args_consumer)
```

Comment:
I wonder if it's possible to unwind these callbacks (e.g. args_consumer)  to make the code more readable. Especially considering that lua doesn't have static typing.

Reply:
There are two possible ways to do that:
1. Replace `mp.command_native_async` with `mp.command_native`. Can block the execution for considerable time (see the timings in the head), but may not be a problem if the rest of MPV remains responsive. Have no idea how MPV handles a single blocked Lua script.
2. Rewrite using Lua coroutines. Never worked with these.

Reply:
What I meant is you could make the args table first, and then execute the args without passing the function that executes them into the function that makes the args.